src/v/storage/kvstore.cc

```diff
@@ -385,7 +385,8 @@ ss::future<> kvstore::recover() {
               [] { return std::nullopt; },
```

Comment:
```
Use `storage::caching_policy` in segment implementation to apply
different caching caching strategies based on the ntp configuration
```

Can you explain how it works? For example I'm looking at

```
    // only caching pending writes, truncate cached batches at new stable
     // offsets as they are no longer needed
     if (_caching_policy == caching_policy::pending_writes) {
         cache_prefix_truncate(it->second);
     }
```

which is invoked by the upcall from the segment appender. it's not clear why we don't leave the entries in the batch cache and let them expire like any other batch. what is the design of caching policies?

also, removing here in the up call for pending write policy implies batches were added as part of the policy (?) but I only see pending_write policy checked once and I was thinking that `full` was not actually a superset of `pending_write` policy. what are the different scenarios that are covered by the change?

Reply:
The problem is that we do not need and want to keep internal topics in cache. They have lots of small batches. This puts a lot of pressure on cache and cache index. We do not need to cache those topics as a whole since they would disturb caching of other topics. Evicting those batches immediately when they are not needed relieves the pressure. 

When it comes to pinning, we agreed some time ago that pinning would need a little bit more thinking since  it will make some of the batches in cache not to be reclaimable.